packages/docusaurus-1.x/lib/static/js/scrollSpy.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/* eslint-disable prefer-arrow-callback */
+(function scrollSpy() {
+  const OFFSET = 10;
+  let timer;
+  let headingsCache;
+  const findHeadings = function findHeadings() {
+    return headingsCache || document.querySelectorAll('.toc-headings > li > a');
+  };
+  const onScroll = function onScroll() {
+    if (timer) {
+      // throttle
+      return;
+    }
+    timer = setTimeout(function() {
+      timer = null;
+      let found = false;
+      const headings = findHeadings();
+      for (let i = 0; i < headings.length; i++) {
+        // if !found and i is the last element, highlight the last
+        let current = !found;
+        if (!found && i < headings.length - 1) {
+          const next = headings[i + 1].href.split('#')[1];
+          const nextHeader = document.getElementById(next);
+          const top = nextHeader.getBoundingClientRect().top;
+          // The following tests whether top + scrollTop
+          // (the top of the header) is greater than scrollTop
+          // (where scrollTop = window.pageYOffset, the top of
+          // the window), with OFFSET pixels of slop.
+          current = top > OFFSET;
+        }
+        if (current) {
+          found = true;
+          headings[i].classList.add('active');
+        } else {
+          headings[i].classList.remove('active');
+        }
+      }
+    }, 100);
+  };
+  document.addEventListener('scroll', onScroll);
+  document.addEventListener('resize', onScroll);
+  document.addEventListener('DOMContentLoaded', function() {
+    // Cache the headings once the page has fully loaded.
+    headingsCache = findHeadings();
+    onScroll();
+    // Find the active nav item in the sidebar
+    const item = document.getElementsByClassName('navListItemActive')[0];
+    if (!item) {
+      return;
+    }
+    const bounding = item.getBoundingClientRect();
+    if (
```

Comment:
I don't really understand what this entire block does, do you mind explaining?

Also it's really uncommon to have an empty `if` block. It'd be better to negate the condition and shift the `else` into the negated `if` condition.

Reply:
Removed unnecessary code from here and also I've left a lot of comments and modified variable names to better understand what is going on. If it still doesn't make any sense I will be happy to explain it here.

Reply:
I'm also trying to learn the chunk, here's my understanding, correct me if I'm wrong, and can consider the following comment?

```javascript
(function scrollSpy() {
  const OFFSET = 10;
  let timer;
  let headingsCache;
  const findHeadings = function findHeadings() {
    return headingsCache || document.querySelectorAll('.toc-headings > li > a');
  };
  const onScroll = function onScroll() {
    if (timer) {
      // throttle
      return;
    }
    /**
     * On every call, try to find header right after  <-- next header
     * the one whose content is on the current screen <-- highlight this
     */
    timer = setTimeout(function() {
      timer = null;
      let activeNavFound = false;
      const headings = findHeadings(); // all of toc nav anchors
      for (let i = 0; i < headings.length; i++) {
        let currNavActive = !activeNavFound;
        /**
         * Enter the following check up only when an active nav header is not yet found
         * Then, check the bounding rectangle of the next header
         * The headers that are scrolled passed will have negative bounding rect top
         * So the first one with positive bounding rect top will be the nearest next header
         */
        if (currNavActive && i < headings.length - 1) {
          const next = headings[i + 1].href.split('#')[1];
          const nextHeader = document.getElementById(next);
          const top = nextHeader.getBoundingClientRect().top;
          currNavActive = top > OFFSET;
        }
        /**
         * Stop searching once a first such header is found,
         * this makes sure the highlighted header is the most current one
         */
        if (currNavActive) {
          activeNavFound = true;
          headings[i].classList.add('active');
        } else {
          headings[i].classList.remove('active');
        }
      }
    }, 100);
  };
```

Reply:
Yupp that's correct. I will update the comments. Thank you :)